### ISSUE TITLE
Async actions

### ISSUE THREAD

**Issue by ivan-kleshnin** [malte-wessel/minimal-flux#1]
Support or example, you know :)

```javascript
async function fetch() {
   return await Axios.get("/some/api")...
}
```

P.S.
Current API looks cute, I like it.

**Comment by malte-wessel** [comment]
Hey Ivan, thanks! 

Completely synchronous means that the API does not have asynchronous methods. Actions can therefore not handle returned Promises. *That was a design decision*. You are maybe familiar with [flummox](https://github.com/acdlite/flummox) where actions can return Promises. The Issue with that is that stores that handle the (async) actions need to take care of the whole flow and intermediate states. With *flummox* you would register an async action by defining a `beginHandler`, `successHandler` and a `failureHandler`. Once the action is invoked, the store itself is responsible for handling the request to your API. I think that this is *beyond the scope of stores*. Actions should take care of anything that goes into or leaves your application. 

With minimal-flux you would do something like this:
````javascript

class Messages extends Actions {
  
  fetch() {
    this.wait();
    Axios.get("/some/api").then(this.complete).catch(this.fail);
  }
  
  wait() {
    this.emit("wait");
  }

  complete(data) {
    this.emit("complete", data);
  }

  fail(err) {
    this.emit("fail", err);
  }
}
````

This is probably a bit more code than just returning a Promise, but you have more control over the flow. You maybe want to inform other parts of your application that you are waiting for a request. Inside the `wait` action you could invoke another action, for example a global loading indicator:

````javascript
wait() {
  // Inside of actions you have access to all other actions that were registered.
  // Maybe there are multiple requests going on and you want show the user a progress bar.
  this.actions.loader.increment();
  this.emit('wait');
}
complete(data) {
  this.actions.loader.decrement();
  this.emit('complete', data);
}
fail(err) {
  this.actions.loader.decrement();
  // If something goes wrong, you maybe want to log it:
  this.actions.logger.add(err);
  this.emit('fail');
}
````

You can find [another example here](https://github.com/malte-wessel/minimal-flux/blob/master/examples/todo/src/actions/TodoActions.js)

Hope that helps :)

**Comment by ivan-kleshnin** [comment]
Thanks. Yes, I'm familiar with flummox. I've been trying about all flux implementations out there :)
What I don't like about flummox is it's API. `let actionIds = flux.getActions("xxx")` is too noisy for me.
You've done a great job by replacing this with properties.

Also your flux == dispatcher. That's good, less concepts to confuse.

You're trying to keep visible and obvious the fact that actions are just event emitters. 
Can't say for sure, whether I like it or not. There are a lot of caused duplications, at least on the first sight. Can action function emit two or more events? Can those events be named differently than function name? I dunno.

It feels cleaner in flummox to just return value or promise instead of emitting them. But they have problems with invoking subsequent actions during the handle of original one (https://github.com/acdlite/flummox/issues/82). Maybe their approach is the straight source of the issue.

Anyway, those string constants (emitted event names) are very easy to put wrong and this will require  debugging (cause no syntax error). That's a pain point.

**Comment by malte-wessel** [comment]
Yes you're right. The emitting part can cause confusion. I'm currently adding some warnings when actions emit events that are not registered as actions. But that shouldn't be the solution. I had another idea for solving this problem:

````javascript
complete(data) {
  // Every action could be bound to an own object that provides 
  // a convenient `dispatch` method. Then you could write:
  this.dispatch(data);
}
````

Under the hood this would call `this.emit('complete', data);`. But one could argue that this is too much magic. I will think about it. 

Thanks for your feedback!

**Comment by ivan-kleshnin** [comment]
I've updated my comment with a note about **Reflux**. I just think that Store and Action should have about the same depth of code layer in this stuff. Now stores have `setState`, but actions use naked `emit`. That's what I found not exactly right.

**Comment by malte-wessel** [comment]
Yes, agree, the code is super clean. But as you pointed out in https://github.com/acdlite/flummox/issues/82 you cannot invoke subsequent actions. Besides that I think there are some problems with invoking actions from stores. This will break the unidirectional data flow, which makes it difficult to reason about changes in state:

````
Component -> Action -> Store -> Action -> Store -> Component
````

What do you mean by that?
> Your implementation of setState is ok, why not apply the same reasoning to actions.?

**Comment by ivan-kleshnin** [comment]
> What do you mean by that?

Same as that:

> I just think that Store and Action should have about the same depth of code layer in this stuff. Now stores use setState, but actions use naked emit. That's what I find not exactly right.

**Comment by ivan-kleshnin** [comment]
> Besides that I think there are some problems with invoking actions from stores. This will break the unidirectional data flow, which makes it difficult to reason about changes in state.

Well I don't really think we have unidirectional data-flow in Flux. We have loop.
`Component > Action > Store > [dependent Stores] > Component`. My "proposal" will add subloops. Without subloops we'll be forced to have additional state variables, or not?

How to add alert on failed load without subsequent action?

**Comment by malte-wessel** [comment]
Ah, ok got it :) Yes, maybe the `dispatch` method for actions is the right way to go. I will check that!

**Comment by malte-wessel** [comment]
Yes, right, kind of a unidirectional loop :) With minimal-flux you could do that by having your alert store listening to any kind of fail action.

````javascript
class Messages extends Actions {
  fetch() {
    Axios.get("/some/api").then(this.complete).catch(this.fail);
  }
  fail(err) {
    this.emit("fail", err);
  }
  // ...
}

class AlertsStore extends Store {
  constructor() {
    this.handleAction('messages.fail', this.addAlert);
  }
  // ...
}
````

That's why I think returning from actions is not the right way.

**Comment by ivan-kleshnin** [comment]
And that makes `AlertStore` dependent of every other alert emitters, as I mentioned... 

```javascript
class AlertsStore extends Store {
  constructor() {
    this.handleAction('users.loadMany.fail', this.addAlert);
    this.handleAction('articles.loadMany.fail', this.addAlert);
    this.handleAction('whatever.loadMany.fail', this.addAlert);
    ...
  }
  // ...
}
```

**Comment by ivan-kleshnin** [comment]
Well it's turns out more interesting than it seems from the first sight.

We can approach this in two ways.

##### Interactive
```javascript
AlertActions < UserStore [calls AlertActions.add()]
AlertActions < DocumentStore [calls AlertActions.add()]
AlertActions < SubscriptionStore [calls AlertActions.add()]
```

##### Reactive
```javascript
UserActions < AlertStore [subsribes to UserActions]
DocumentActions < AlertStore [subsribes to DocumentActions]
SubscriptionActions < AlertStore [subsribes to SubscriptionActions]
```

Replacing with letters for brevity and removing Store vs Action suffix which doesn't matter...:

##### Interactive
```javascript
A < U
A < D
A < S 
```

* `A` can be library, `U`, `S`, `D` - can't
* Dependencies from `A` are scattered
* Changes in `A` may break `U`, `D`, `C` 

##### Reactive
```javascript
U < A
D < A
S < A
```

* `U`, `D`, `S` can be library, `A` - can't
* `A` module contains all it's dependencies
* Changes in `U`, `D`, `C` may break `A`

I bet there's more info I can't extract right now...
